### DISCUSSION TITLE
Mill Docker Jib Plugin

### DISCUSSION THREAD

**Discussion by GeorgOfenbeck** [com-lihaoyi/mill#3900] [Show and tell]
Hi Mill Community,

I'm a big fan of mill and been using it for my private toy projects since a while now. While I don't feel its yet managable by non-scala experienced users nor stable enough (which is the reason I never introduced it in my work setting) I definitely hope that it will be soon.
Having to deal with Gradle, Maven and even Ant (don't ask 🙈) always makes it a delight working with Mill again.

As a first step to try to contribute to Mills future I tried myself on writing a mill plugin to learn more mill internals. Hopefully as a first step to contribute more later...

# The plugin

The plugin can be found here:
[Mill Docker Jib](https://github.com/GeorgOfenbeck/mill-docker)

I'm still working on it - particularly getting some more testing in and the items listed under future work.
Happy to get any feedback.

If it's ok I would make a pull request to add it to the [Third Party Plugins](https://mill-build.org/mill/0.12.1/extending/thirdparty-plugins.html) ?
Might also be worth mentioning the alternatives (native image, this one etc) on the contrib page?

# Feedback / Questions for Mill

## Itest vs Testkit
I assume the [mill-integrationtest Plugin](https://github.com/lefou/mill-integrationtest) is obsolete with 0.12.x and the introduction of testkit?
Maybe a remark on the repo would be good @lefou - i started with your great plugin and didn't immediately figure out about the new built in testkit

## Docu broken link
~~https://mill-build.org/mill/0.11.12/index.html (From the dropdown)~~ already fixed

## T macro stuff
I ran into various smaller issues - one I discussed with @lihaoyi on discord. The beauty of Mill for me is how easy it is to reproduce whats going on with the File based build output and the dependency graph.

The Task macros break that feeling as there suddenly is "magic".
Of course one can figure it out eventually as a scala veteran - but I'm afraid for non-Scala or beginner Scala users certain macro errors will definitely be very frustrating.

The simplest "gotcha" is just type inference e.g.
`def foo: T[Seq[String]] = Seq.empty` 

The bracket vs no brackets on e.g. Commands (its documented in the docs)
E.g. it took me a moment to realize when you need the extra brackets in e.g.:
`
def println1() = T.command{ println("Hello World") }
def println2() = T.command{ println("Hello World 2") }
def both() = T.command{ 
  val x = println1()(); 
  val y = println2()();
  x 
}
`

In my plugin I had the case where I wanted to pass a function around - that caused all sorts of issues. A minimal version of the issue can be seen [here](https://gist.github.com/GeorgOfenbeck/6a1c562daa5a8fa8ebd8b256175966c7). In the end I managed - but had to do a strange tail calling like [here](https://github.com/GeorgOfenbeck/mill-docker/blob/7dd39d1a108803004361b5073c08d8bf95d7c23b/mill-docker/src/com/ofenbeck/mill/docker/DockerJibModule.scala#L198)

Again - all solvable - but I feel thats currently the biggest "magic" part of mill. 
As such I'm curious to see how this changes with [#3369](https://github.com/com-lihaoyi/mill/pull/3369)

# Scalafmt
I switched to .mill file extension - might be worth mentioning that one needs to add it to the file list at the moment. (or upgrade to latest RC of scalafmt)

# Cross
~~When running a test on e.g [0.12.1] it will still compile [0.12.0] assuming both a part of the cross. (compile both - test only the one specified)~~
Edit - resolved: Was using `mill-docker`() which resolved to the default cross element and as such i got this dependency "across" the cross

# Testkit
## Example Tester
Would be nice if the output was somehow visible in he "non-error" case. E.g in a log file. 
Of course this is only for plugin writers who will figure out other ways - but was searching a bit at first to see if i can get the output somewhere.

The windows hardcoded paths are a bit iffy - then again I guess the chances of someone developing a plugin in pure windows are close to zero :P

Having an out folder in the Example project breaks the test (e.g. actually running the example by hand)

# Remarks
- Publishing with the _mill0.11 was a bit confusing at first - guess platform authors need to put compatibility matrices on their page like @lefou did for his plugins.
- ran into the zsh wildcard vs mill wildcard issue - realized only latter that its in the docu

**Comment by lefou** [comment]
Welcome to the Mill community and ecosystem! Please tag your Github project with `mill-plugin`, so it can be found easily. Feel free to send PRs! If you unsure open an issue or make the PR a draft first. There is no need to ask, we already invited anybody to send PRs to add plugins to the Third Party Plugins list. I'm looking forward to it.

I really appreciate your feedback. Yet, to keep focus and improve over time it's better to open separate issues, so we don't loose track and can discuss each issue separately.

Regarding the mill-integrationtest project. I haven't found the time to test, whether it is still workng with Mill 0.12. It should, but if not, I intend to keep it working for a while (PRs welcome). Aas long as there are plugins build against older Mill versions, there is no replacement in Mill Testkit. Generally, I think Mill Testkit is great alternative. I added some verbose output in the past, but it somehow vanished in later changes. I didn't find the time to look into that. But having a verbose mode, maybe enabled via a Java system property, would be nice.

**Reply by GeorgOfenbeck** [reply]
Oki thanks Tobias - for the things that can be fixed/reported i'll create the according PR's and issues in the coming days.
For the T[_] macro stuff I'll try to follow the [Scala3 effort](https://github.com/com-lihaoyi/mill/pull/3369) to see how the issue changes.

For the mill-integration test:
I had some small troubles with 0.12 - just mention them here 
- at the time i worked with it 0.12 was still RC and as such the download didnt work - could be fixed by manually putting the mill version in the .cache. (snot an issue any more with official release)
- the $file.plugin mechanism with 0.12.x seemded not to work - i did not investigate and instead just published locally and used the regular ivy import.

As I now switched to testkit I wont investigate further, so just a heads up here - I wont create an issue for that one as its a bit hand wavy  and not properly verified

**Reply by lefou** [reply]
Thank you for the feedback to mill-integrationtest. I'll look into it.
